File: applications/brill_wave/src/conformal_solver.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, Optional, Tuple

import numpy as np


@dataclass(frozen=True)
class ConvergenceInfo:
    iterations: int
    converged: bool
    last_error: float


class ConformalFactorSolver:
    """Iterative solver for the conformal factor psi = 1 + chi.

    This follows the algorithm you pasted. Where the algorithm says “using FMM”,
    this implementation uses a brute-force particle-particle summation.

    Notes
    -----
    - Points are treated as “particles” located at r_a with associated volumes V_a.
    - The brute-force kernel is O(N^2) per iteration and is intended for small N.
    - A small softening length can be used to avoid singular self-interactions.
    """

    def __init__(
        self,
        r: np.ndarray,
        V: np.ndarray,
        volumes: np.ndarray,
        *,
        M: float,
        softening: float = 0.0,
        exclude_self: bool = True,
        dtype=np.float64,
    ) -> None:
        r = np.asarray(r, dtype=dtype)
        V = np.asarray(V, dtype=dtype)
        volumes = np.asarray(volumes, dtype=dtype)

        if r.ndim != 2 or r.shape[1] != 3:
            raise ValueError("r must have shape (N, 3)")
        if V.shape != (r.shape[0],):
            raise ValueError("V must have shape (N,)")
        if volumes.shape != (r.shape[0],):
            raise ValueError("volumes must have shape (N,)")
        if M <= 0:
            raise ValueError("M must be positive")
        if softening < 0:
            raise ValueError("softening must be non-negative")

        self.r = r
        self.V = V
        self.volumes = volumes
        self.M = float(M)
        self.softening = float(softening)
        self.exclude_self = bool(exclude_self)

        self._W = float(np.sum(self.volumes * self.V))

    @property
    def W(self) -> float:
        """Discrete approximation W = ∫ V d^3r."""
        return self._W

    def initial_guess_tilde_chi(self) -> np.ndarray:
        """Default initial guess: \tilde{chi}^{(0)} = (1 + 4 r^2 / M^2)^(-1/2)."""
        r2 = np.sum(self.r * self.r, axis=1)
        return 1.0 / np.sqrt(1.0 + 4.0 * r2 / (self.M * self.M))

    def normalize_chi(self, tilde_chi: np.ndarray) -> Tuple[float, np.ndarray]:
        """Step 6: normalize using the discrete version of the paper's equation."""
        tilde_chi = np.asarray(tilde_chi, dtype=self.r.dtype)
        if tilde_chi.shape != (self.r.shape[0],):
            raise ValueError("tilde_chi must have shape (N,)")

        denom = float(np.sum(self.volumes * self.V * tilde_chi))
        if denom == 0.0:
            raise ZeroDivisionError(
                "Normalization denominator is zero: sum(volumes * V * tilde_chi)"
            )

        C = -(4.0 * np.pi * self.M + self.W) / denom
        chi = C * tilde_chi
        return float(C), chi

    def masses_from_chi(self, chi: np.ndarray) -> np.ndarray:
        """Step 7: m_a = volumes_a * V_a * (1 + chi_a) / (4π)."""
        chi = np.asarray(chi, dtype=self.r.dtype)
        if chi.shape != (self.r.shape[0],):
            raise ValueError("chi must have shape (N,)")
        return (self.volumes * self.V * (1.0 + chi)) / (4.0 * np.pi)

    def _pairwise_kernel(self, targets: np.ndarray, sources: np.ndarray) -> np.ndarray:
        """Return 1/|x_i - y_j| with optional softening."""
        diff = targets[:, None, :] - sources[None, :, :]
        d2 = np.sum(diff * diff, axis=2)
        if self.softening > 0.0:
            d2 = d2 + self.softening * self.softening
        return 1.0 / np.sqrt(d2)

    def chi_from_masses(self, m: np.ndarray) -> np.ndarray:
        """Brute-force version of step 9.

        Uses the same kernel implied by step 12:
            psi(r) = 1 - (1/4π) * Σ_a m_a / |r - r_a|
        hence
            chi(r) = psi - 1 = - (1/4π) * Σ_a m_a / |r - r_a|.

        For points exactly coincident with a source, this implementation either
        excludes self-interactions (default) or uses a softening length.
        """
        m = np.asarray(m, dtype=self.r.dtype)
        if m.shape != (self.r.shape[0],):
            raise ValueError("m must have shape (N,)")

        inv_r = self._pairwise_kernel(self.r, self.r)
        if self.exclude_self and self.softening == 0.0:
            np.fill_diagonal(inv_r, 0.0)

        potential = inv_r @ m
        return -(1.0 / (4.0 * np.pi)) * potential

    def solve(
        self,
        *,
        tol: float = 1e-10,
        max_iter: int = 200,
        error_norm: Literal["linf", "l2"] = "linf",
        verbose: bool = False,
        initial_tilde_chi: Optional[np.ndarray] = None,
    ) -> Tuple[np.ndarray, ConvergenceInfo]:
        """Run steps 4-11 on the interpolation points.

        Returns
        -------
        psi : ndarray, shape (N,)
            Conformal factor at the interpolation points.
        info : ConvergenceInfo
            Iteration diagnostics.
        """
        if tol <= 0:
            raise ValueError("tol must be positive")
        if max_iter <= 0:
            raise ValueError("max_iter must be positive")

        if initial_tilde_chi is None:
            tilde_chi = self.initial_guess_tilde_chi()
        else:
            tilde_chi = np.asarray(initial_tilde_chi, dtype=self.r.dtype)

        last_err = np.inf
        for n in range(max_iter):
            _, chi = self.normalize_chi(tilde_chi)
            m = self.masses_from_chi(chi)

            tilde_next = self.chi_from_masses(m)

            delta = tilde_next - tilde_chi
            if error_norm == "linf":
                err = float(np.max(np.abs(delta)))
            elif error_norm == "l2":
                err = float(np.linalg.norm(delta) / np.sqrt(delta.size))
            else:
                raise ValueError("error_norm must be 'linf' or 'l2'")

            if verbose:
                print(f"iter {n:4d}  err={err:.3e}")

            tilde_chi = tilde_next
            last_err = err
            if err < tol:
                _, chi = self.normalize_chi(tilde_chi)
                psi = 1.0 + chi
                return psi, ConvergenceInfo(iterations=n + 1, converged=True, last_error=err)

        # not converged
        _, chi = self.normalize_chi(tilde_chi)
        psi = 1.0 + chi
        return psi, ConvergenceInfo(iterations=max_iter, converged=False, last_error=last_err)

    def psi_at_points(self, r_eval: np.ndarray, *, chi: np.ndarray) -> np.ndarray:
        """Step 12: evaluate psi(r) at arbitrary points via brute-force summation."""
        r_eval = np.asarray(r_eval, dtype=self.r.dtype)
        if r_eval.ndim != 2 or r_eval.shape[1] != 3:
            raise ValueError("r_eval must have shape (K, 3)")

        m = self.masses_from_chi(chi)
        inv_r = self._pairwise_kernel(r_eval, self.r)

        potential = inv_r @ m
        return 1.0 - (1.0 / (4.0 * np.pi)) * potential

File: applications/brill_wave/c/src/conformal_solver.cpp
#include "conformal_solver.hpp"

#include <cassert>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

#ifndef M_PI
#define M_PI 3.141592653589793238462643383279502884
#endif

//static double cs_sq(double x) { return x * x; }

//void conformal_solver_t::init(size_t n_, real M_, real softening_,
//							  bool exclude_self_) {
//	n = n_;
//	r_xyz.resize(n_);
//	V.resize(n);
//	vol.resize(n);
//
//	assert(n > 0);
//	assert(M > 0.0);
//	assert(softening > 0.0);
//
//	r_xyz.resize(n);
//	V.resize(n);
//	vol.resize(n);
//	M = M_;
//	softening = softening_;
//	exclude_self = exclude_self_ ? 1 : 0;
//}
//
//void cs_initial_guess_tilde_chi(const conformal_solver_t *s,
//								double *tilde_chi_out) {
//	const size_t n = s->n;
//	const double invM2 = 1.0 / (s->M * s->M);
//
//	for (size_t i = 0; i < n; ++i) {
//		const double x = s->r_xyz[3 * i + 0];
//		const double y = s->r_xyz[3 * i + 1];
//		const double z = s->r_xyz[3 * i + 2];
//		const double r2 = x * x + y * y + z * z;
//		tilde_chi_out[i] = 1.0 / sqrt(1.0 + 4.0 * r2 * invM2);
//	}
//}
//
//int cs_normalize_chi(const conformal_solver_t *s, const double *tilde_chi,
//					 double *C_out, double *chi_out) {
//	const size_t n = s->n;
//	double denom = 0.0;
//	for (size_t i = 0; i < n; ++i) {
//		denom += s->vol[i] * s->V[i] * tilde_chi[i];
//	}
//	if (denom == 0.0)
//		return 1;
//
//	const double C = -(4.0 * M_PI * s->M + s->W) / denom;
//	if (C_out)
//		*C_out = C;
//	for (size_t i = 0; i < n; ++i) {
//		chi_out[i] = C * tilde_chi[i];
//	}
//	return 0;
//}
//
//void cs_masses_from_chi(const conformal_solver_t *s, const double *chi,
//						double *m_out) {
//	const size_t n = s->n;
//	const double inv4pi = 1.0 / (4.0 * M_PI);
//	for (size_t i = 0; i < n; ++i) {
//		m_out[i] = s->vol[i] * s->V[i] * (1.0 + chi[i]) * inv4pi;
//	}
//}
//
//void cs_chi_from_masses(const conformal_solver_t *s, const double *m,
//						double *tilde_chi_next_out) {
//	const size_t n = s->n;
//	const double inv4pi = 1.0 / (4.0 * M_PI);
//	const double eps2 = s->softening * s->softening;
//
//	for (size_t i = 0; i < n; ++i) {
//		const double xi = s->r_xyz[3 * i + 0];
//		const double yi = s->r_xyz[3 * i + 1];
//		const double zi = s->r_xyz[3 * i + 2];
//
//		double sum = 0.0;
//		for (size_t j = 0; j < n; ++j) {
//			if (s->exclude_self && eps2 == 0.0 && j == i)
//				continue;
//
//			const double xj = s->r_xyz[3 * j + 0];
//			const double yj = s->r_xyz[3 * j + 1];
//			const double zj = s->r_xyz[3 * j + 2];
//
//			const double dx = xi - xj;
//			const double dy = yi - yj;
//			const double dz = zi - zj;
//			const double r2 = dx * dx + dy * dy + dz * dz + eps2;
//			sum += m[j] / sqrt(r2);
//		}
//
//		tilde_chi_next_out[i] = -inv4pi * sum;
//	}
//}
//
//static double cs_error_linf(size_t n, const double *a, const double *b) {
//	double m = 0.0;
//	for (size_t i = 0; i < n; ++i) {
//		const double d = fabs(a[i] - b[i]);
//		if (d > m)
//			m = d;
//	}
//	return m;
//}
//
//static double cs_error_l2(size_t n, const double *a, const double *b) {
//	double s2 = 0.0;
//	for (size_t i = 0; i < n; ++i) {
//		const double d = a[i] - b[i];
//		s2 += d * d;
//	}
//	return sqrt(s2 / (double)n);
//}
//
//int cs_solve(const conformal_solver_t *s, double tol, int max_iter,
//			 cs_error_norm_t norm, int verbose, double *psi_out,
//			 cs_convergence_info_t *info_out) {
//	if (!s || !psi_out || !info_out)
//		return 1;
//	if (!(tol > 0.0))
//		return 2;
//	if (max_iter <= 0)
//		return 3;
//
//	const size_t n = s->n;
//
//	double *tilde = (double *)malloc(n * sizeof(double));
//	double *tilde_next = (double *)malloc(n * sizeof(double));
//	double *chi = (double *)malloc(n * sizeof(double));
//	double *m = (double *)malloc(n * sizeof(double));
//
//	if (!tilde || !tilde_next || !chi || !m) {
//		free(tilde);
//		free(tilde_next);
//		free(chi);
//		free(m);
//		return 4;
//	}
//
//	cs_initial_guess_tilde_chi(s, tilde);
//
//	double last_err = INFINITY;
//	int converged = 0;
//	int iters = 0;
//
//	for (int iter = 0; iter < max_iter; ++iter) {
//		double C = 0.0;
//		if (cs_normalize_chi(s, tilde, &C, chi) != 0) {
//			free(tilde);
//			free(tilde_next);
//			free(chi);
//			free(m);
//			return 5;
//		}
//		cs_masses_from_chi(s, chi, m);
//		cs_chi_from_masses(s, m, tilde_next);
//
//		double err = 0.0;
//		if (norm == CS_ERROR_NORM_LINF) {
//			err = cs_error_linf(n, tilde_next, tilde);
//		} else {
//			err = cs_error_l2(n, tilde_next, tilde);
//		}
//
//		if (verbose) {
//			printf("iter %4d  err=%.3e\n", iter, err);
//		}
//
//		/* swap tilde buffers */
//		double *tmp = tilde;
//		tilde = tilde_next;
//		tilde_next = tmp;
//
//		last_err = err;
//		iters = iter + 1;
//		if (err < tol) {
//			converged = 1;
//			break;
//		}
//	}
//
//	/* final normalization for output psi */
//	if (cs_normalize_chi(s, tilde, NULL, chi) != 0) {
//		free(tilde);
//		free(tilde_next);
//		free(chi);
//		free(m);
//		return 6;
//	}
//	for (size_t i = 0; i < n; ++i) {
//		psi_out[i] = 1.0 + chi[i];
//	}
//
//	info_out->iterations = iters;
//	info_out->converged = converged;
//	info_out->last_error = last_err;
//
//	free(tilde);
//	free(tilde_next);
//	free(chi);
//	free(m);
//	return 0;
//}

//void cs_psi_at_points(const conformal_solver_t *s, const double *chi, size_t k,
//					  const double *r_eval_xyz, double *psi_out) {
//	const size_t n = s->n;
//	const double inv4pi = 1.0 / (4.0 * M_PI);
//	const double eps2 = s->softening * s->softening;
//
//	/* Precompute masses from chi */
//	double *m = (double *)malloc(n * sizeof(double));
//	if (!m)
//		return;
//	cs_masses_from_chi(s, chi, m);
//
//	for (size_t i = 0; i < k; ++i) {
//		const double xi = r_eval_xyz[3 * i + 0];
//		const double yi = r_eval_xyz[3 * i + 1];
//		const double zi = r_eval_xyz[3 * i + 2];
//
//		double sum = 0.0;
//		for (size_t j = 0; j < n; ++j) {
//			const double xj = s->r_xyz[3 * j + 0];
//			const double yj = s->r_xyz[3 * j + 1];
//			const double zj = s->r_xyz[3 * j + 2];
//			const double dx = xi - xj;
//			const double dy = yi - yj;
//			const double dz = zi - zj;
//			const double r2 = dx * dx + dy * dy + dz * dz + eps2;
//			sum += m[j] / sqrt(r2);
//		}
//		psi_out[i] = 1.0 - inv4pi * sum;
//	}
//
//	free(m);
//}
